### ISSUE TITLE
ErrorException: Warning: PDO::prepare(): MySQL server has gone away

### ISSUE THREAD

**Issue by keithbrink** [laravel/horizon#583]
- Horizon Version: 3.0.6
- Laravel Version: 5.7.28
- PHP Version: 7.2.17
- Redis Driver & Version: predis/phpredis 1.1.1
- Database Driver & Version: MariaDB 10.2.23

### Description:

I am struggling with a bug on my production server using Horizon.

ErrorException: Warning: PDO::prepare(): MySQL server has gone away [internal] in unserialize

You can see a stack trace of the error here: https://sentry.io/share/issue/b105b7946b524a9e841f56f44445ea14/

As far as I can tell, this error should be caught by the Laravel framework. I'm not sure why it's not being caught and turned into a QueryException which would then trigger the reconnection and/or killing the worker.

See: https://github.com/laravel/framework/blob/9fb420cc29a7dd5de5051f09c523ffc3ea01b969/src/Illuminate/Database/Connection.php#L663

And then: https://github.com/laravel/framework/blob/9fb420cc29a7dd5de5051f09c523ffc3ea01b969/src/Illuminate/Database/Connection.php#L735

My understanding is that any Exception should be caught and then re-thrown as a QueryException, which would then be properly caught by the framework and then reconnected to the database.

This is an occasional error so it's difficult to reproduce; I've tried to manually throw a similar error but it is caught properly and handled properly.

Any general guidance on why this error might be different in production and ideas on how I can isolate the error would be appreciated.

**Comment by driesvints** [comment]
Hey there,

Can you first please try one of the following support channels? If you can actually identify this as a bug, feel free to report back and I'll gladly help you out.

- [Laracasts Forums](https://laracasts.com/discuss)
- [Laravel.io Forums](https://laravel.io/forum)
- [StackOverflow](https://stackoverflow.com/questions/tagged/laravel)
- [Discord](https://discordapp.com/invite/KxwQuKb)
- [Larachat](https://larachat.co)
- [IRC](https://webchat.freenode.net/?nick=laravelnewbie&channels=%23laravel&prompt=1)

Thanks!

**Comment by drbyte** [comment]
@keithbrink I've been seeing the same thing intermittently for several weeks. 
laravel/framework                        v5.8.14
laravel/horizon                          v3.1.1      
doctrine/dbal                            v2.9.2     
predis/predis                            v1.1.1       
PHP 7.3.4 on a Forge server

At first I thought it might be related to devops things where the server was being updated, but I've confirmed that that's not the case.

As you say, it's hard to reproduce. I've tried similar things as you described.

Horizon doesn't report it as a failed job. I only discover it in my Sentry logs.

Here are my logs -- pretty much identical to yours except for the line numbers (due to laravel version differences), and I've got doctrine/dbal installed which is why it is mentioned here.
```
ErrorException: Warning: PDO::prepare(): MySQL server has gone away
#42 /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php(61): Doctrine\DBAL\Driver\PDOConnection::prepare
#41 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(326): Illuminate\Database\Connection::Illuminate\Database\{closure}
#40 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(657): Illuminate\Database\Connection::runQueryCallback
#39 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(624): Illuminate\Database\Connection::run
#38 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(333): Illuminate\Database\Connection::select
#37 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2124): Illuminate\Database\Query\Builder::runSelect
#36 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2112): Illuminate\Database\Query\Builder::Illuminate\Database\Query\{closure}
#35 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2598): Illuminate\Database\Query\Builder::onceWithColumns
#34 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2113): Illuminate\Database\Query\Builder::get
#33 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(519): Illuminate\Database\Eloquent\Builder::getModels
#32 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(503): Illuminate\Database\Eloquent\Builder::get
#31 /app/Jobs/SendSignupAcknowledgements.php(36): App\Jobs\SendSignupAcknowledgements::handle
#30 [internal](0): call_user_func_array
#29 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}
#28 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\Container\BoundMethod::callBoundMethod
#27 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\Container\BoundMethod::call
#26 /vendor/laravel/framework/src/Illuminate/Container/Container.php(576): Illuminate\Container\Container::call
#25 /vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(94): Illuminate\Bus\Dispatcher::Illuminate\Bus\{closure}
#24 /vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
#23 /vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\Pipeline\Pipeline::then
#22 /vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(98): Illuminate\Bus\Dispatcher::dispatchNow
#21 /vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(49): Illuminate\Queue\CallQueuedHandler::call
#20 /vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(86): Illuminate\Queue\Jobs\Job::fire
#19 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(327): Illuminate\Queue\Worker::process
#18 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(277): Illuminate\Queue\Worker::runJob
#17 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(118): Illuminate\Queue\Worker::daemon
#16 /vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(102): Illuminate\Queue\Console\WorkCommand::runWorker
#15 /vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(86): Illuminate\Queue\Console\WorkCommand::handle
#14 /vendor/laravel/horizon/src/Console/WorkCommand.php(46): Laravel\Horizon\Console\WorkCommand::handle
#13 [internal](0): call_user_func_array
#12 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}
#11 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\Container\BoundMethod::callBoundMethod
#10 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\Container\BoundMethod::call
#9 /vendor/laravel/framework/src/Illuminate/Container/Container.php(576): Illuminate\Container\Container::call
#8 /vendor/laravel/framework/src/Illuminate/Console/Command.php(183): Illuminate\Console\Command::execute
#7 /vendor/symfony/console/Command/Command.php(255): Symfony\Component\Console\Command\Command::run
#6 /vendor/laravel/framework/src/Illuminate/Console/Command.php(170): Illuminate\Console\Command::run
#5 /vendor/symfony/console/Application.php(908): Symfony\Component\Console\Application::doRunCommand
#4 /vendor/symfony/console/Application.php(269): Symfony\Component\Console\Application::doRun
#3 /vendor/symfony/console/Application.php(145): Symfony\Component\Console\Application::run
#2 /vendor/laravel/framework/src/Illuminate/Console/Application.php(90): Illuminate\Console\Application::run
#1 /vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(122): Illuminate\Foundation\Console\Kernel::handle
#0 /artisan(37): null
```

**Comment by keithbrink** [comment]
@drbyte Thanks for your feedback, and I will report back if I make any progress on this.

I posted this to StackOverflow last week but no takers so far: https://stackoverflow.com/questions/55878253/laravel-horizon-errorexception-warning-pdoprepare-mysql-server-has-gone

**Comment by driesvints** [comment]
Heya you both. Feel free to reply after monitoring a few days if this continuous to happen and I'll try to help out.

**Comment by keithbrink** [comment]
@drbyte @driesvints Is it possible that Sentry is logging this error but the worker is being restarted properly? None of these errors are showing up in the failed jobs of Horizon, so I wonder if Sentry is incorrectly catching these errors?

**Comment by driesvints** [comment]
Could be the case yeah. Does the result of whatever `SendSignupAcknowledgements` work out properly?

**Comment by ikudosi** [comment]
I also have the same issue (if it is even an issue). I have a job running twice a day and all it's doing is sending out a simple email with a URL. Sentry has logged it as an error and Horizon has no recollection of a failed job. However... the job did process successfully as the end users were able to get the email.

**Comment by sebestenyb** [comment]
We started seeing this error since we upgraded to Laravel 5.8, pretty much every `horizon:work` ends up with this error in Sentry.

**Comment by jlmmns** [comment]
Same thing over here.
Upgrade to 5.8. Everything seems to work fine, but Sentry keeps logging those exceptions.

**Comment by keithbrink** [comment]
@driesvints Any ideas for how we can get rid of the error from Sentry since the jobs are completing successfully?

**Comment by driesvints** [comment]
No idea sorry...

**Comment by drbyte** [comment]
Yes, still happening for me. 
I'm pretty sure the jobs actually do perform correctly.

Adding my updated version info for context:
`Laravel 5.8.28`
`mysql  Ver 15.1 Distrib 10.2.25-MariaDB` and `10.1.36-MariaDB`
Happens with and without `telescope` installed.
Horizon is of course using `redis` for the queue.

I'm seeing both the `STMT_PREPARE` and `MySQL server has gone away` errors (see linked issues above).

**Comment by gpn273** [comment]
This happening for me. I am assuming the issue is caused because of the long-running command (`php artisan horizon`) has surpassed it's maximum connection time for the database and once a job has been received onto Redis, the job attempts to do a database operation but is faced with the following warning: `Warning: Error while sending STMT_PREPARE packet. PID=22312`.

My current production versions:
laravel/framework `v6.0.3`
laravel/horizon `v3.3.2`

Does anyone know of any possible solutions/work arounds?

**Comment by goyote** [comment]
Can you guys post you finding?

**Comment by keithbrink** [comment]
Fixed according to this comment: https://github.com/getsentry/sentry-laravel/issues/254#issuecomment-543558493

**Comment by jaimyborgman** [comment]
I've posted my findings a few months ago in another thread here: https://github.com/laravel/framework/issues/15179#issuecomment-519073609, but still having this issue. In my case it's not really something related to sentry. It also took place after updating probably from 5.7 to 5.8 as reading on several other threads as well but however i'm not sure about this.

In the meantime after posting my previous message i've been doing some additional research, updated mysql configuration to get the best setup as possible, updated dependencies and much more but still having this issue as of today.

I know that in my case it has nothing to do with Sentry, altho jobs are marked as completed and finished "correctly" the outcome of such a job is incorrect. Like in my case for example it is checking whether to send new invoice notifications to a user, while sending it will check if we could collect the payment automatically or not based on a flag in the database it will set an additional message in the email send to the client (we are collecting or manual payment needs to be done).

In this example the client has actually a valid "mandate" to do auto collecting of payments, but however while sending that notification it's trying to connect to the database, will get the occasional message like "server has gone away" or "error while sending query packet" and then ending up sending the message without a valid mandate since it couldn't retrieve that info from the database.

So in short; i don't think it has anything to do with Sentry since it's logging those errors correctly as since they occur in the system. It more looks like something related to the laravel framework or laravel-doctrine.

**Comment by keithbrink** [comment]
@jaimyborgman Are you on sentry-laravel 1.4.0?

**Comment by jaimyborgman** [comment]
@keithbrink i've updated everything yesterday to make sure i'm on the latest versions, it looked liked the errors where at least gone for a few hours but overnight i've got them again: "PDOException: SQLSTATE[HY000]: General error: 2006 MySQL server has gone away" and "ErrorException: Error while sending QUERY packet. PID=*".

Currently installed versions are:
PHP (7.2.22)
MySQL (5.7.27)
laravel/framework (v5.8.35)
doctrine/orm (v2.6.4)
laravel-doctrine/orm (1.4.11)
sentry/sentry (2.2.2)
sentry/sentry-laravel (1.4.1)

**Comment by mrb235** [comment]
We've been having this issue for the last couple days and we finally figured it out.  It looks like the doctrine connection isn't closed by the horizon jobs and then they hang out until it gets timed out by the database.  We resolved this by adding event handlers to the AppServiceProvider that close the connection after successful and failed jobs.  I'm not sure if there is a cleaner way to accomplish this, but this does work.  
```        
    public function boot() {
        Queue::failing(function(JobFailed $event) {
            $entityManager = new EntityManager();
            $entityManager->getConnection()->close()
        });
        Queue::after(function(JobProcessed $event) {
            $entityManager = new EntityManager();
            $entityManager->getConnection()->close()
        });
    }
```

**Comment by jaimyborgman** [comment]
That was exactly what i was thinking too, i've checked hardware and other points as well in the meantime but all was pointing to horizon. So i will check this solution as well; what i also saw was the amount of aborted client connections was pretty high. It looks like something in an package update changed some logic.

Many thanks for pointing into this direction @mrb235 i will test and try on my machines as well and will confirm if this was a solution here as well (i think it does).

**Comment by keithbrink** [comment]
> We've been having this issue for the last couple days and we finally figured it out. It looks like the doctrine connection isn't closed by the horizon jobs and then they hang out until it gets timed out by the database. We resolved this by adding event handlers to the AppServiceProvider that close the connection after successful and failed jobs. I'm not sure if there is a cleaner way to accomplish this, but this does work.
> 
> ```
>     public function boot() {
>         Queue::failing(function(JobFailed $event) {
>             $entityManager = new EntityManager();
>             $entityManager->getConnection()->close()
>         });
>         Queue::after(function(JobProcessed $event) {
>             $entityManager = new EntityManager();
>             $entityManager->getConnection()->close()
>         });
>     }
> ```

@driesvints Is this something that could get added to the Laravel core?

**Comment by jaimyborgman** [comment]
It is working; i've tweaked it a little tho for my scenario but not much only where it's getting its entitymanager from (in my case from the container instead of making a new one) but to be honest i don't know if this should be part of the Laravel core; i think it should be find out by what this issue occurred because before an update it was perfectly fine.

`if (! $this->app->runningUnitTests()) {
            Queue::after(function (JobProcessed $event) {
                /** @var EntityManager $entityManager */
                $entityManager = $this->app['em'];
                $entityManager->getConnection()->close();
            });

            Queue::failing(function (JobFailed $event) {
                /** @var EntityManager $entityManager */
                $entityManager = $this->app['em'];
                $entityManager->getConnection()->close();
            });
        }`

i've placed `! $this->app->runningUnitTests()` on purpose since it works on my staging/production environment but locally for testing it failed all tests

**Comment by fhferreira** [comment]
```
laravel/framework                      v6.20.44               The Laravel Framework.
laravel/horizon                           v3.7.2                 Dashboard and code-driven configuration for Laravel queues.
sentry/sentry-laravel                 4.1.1                  Laravel SDK for Sentry (https://sentry.io)
```

**SQLSTATE[HY000]: General error: 2006 MySQL server has gone away (SQL: select * from `table_x` where `id` = 123 limit 1)**

Here apparently I have the same false positive with laravel-sentry.
